### PR TITLE
disconnected clients: ensure servers meet minimum required version

### PR DESCRIPTION
**Pull request by DerekStrickland**
**Note for Reviewer**: This is a pragmatic compromise that might need discussion. 
- The `CoreScheduler` has a direct reference to the `nomad.Server` and thus its members and is responsible for 
  performing server version checks for core jobs.
- Initially, I tried exposing this functionality via the `Scheduler` interface, however, the `GenericScheduler` cannot have a 
  direct reference to the `nomad.Server` because it is in the `scheduler` package and that creates an import cycle.
- The `Worker` fulfills the `Planner` interface, is made available to the `GenericScheduler` by that interface, and also 
  has a reference to the `nomad.Server`. So the `Planner` interface was extended instead.

## Overview 
This PR contains the logic necessary to ensure clusters running mixed version servers do not enable the `max_client_disconnect` feature until all servers meet the minimum required version. 
- [bf71504](https://github.com/hashicorp/nomad/pull/12202/commits/bf71504d0158d622059f506e88974e412f5843bc)
  - Exposes the `ServersMeetMinimumVersion` utility function via the `Planner` interface. 
- [6e60a2c](https://github.com/hashicorp/nomad/pull/12202/commits/6e60a2c9285d1c47788fcc3285e3596345297732) 
  - Modifies `filterByTained` to include a boolean flag indicating whether the caller supports disconnected clients. 
  - If not, the `disconnected` and `reconnecting` `allocSet`s are never populated.
  - Effectively short-circuits disconnect/reconnect logic from ever triggering in the `reconciler`.  
- [b41401b](https://github.com/hashicorp/nomad/pull/12202/commits/b41401b4fece2d94e63b608da20658672a8f542f) 
  - Adds the version check logic in the `GenericScheduler`
  - Adds a field to the `allocReconciler`.
  - Modifies`NewAllocReconciler` to accept a parameter to set the field.
  - Plumbs it through to all calls to`filterByTainted`. 
- [cfc5f49](https://github.com/hashicorp/nomad/pull/12202/commits/cfc5f49c071524ecc1c901506b25678904d3b644)
  - Updates affected/dependent tests

### CODE REVIEW

File: scheduler/generic_sched.go
Comment (DerekStrickland) [inline scheduler/generic_sched.go:73]: @tgross Should I bump the version number in my branch to `1.3.0` or do we have a technique for faking this during development?

Comment (tgross) [reply scheduler/generic_sched.go:73]: Tricky... we don't bump `version.go` until the actual release so even if you swapped out your feature branch to 1.3.0 it'd be broken once we merged it to `main`. But the version we're developing on isn't `1.2.6`, it's `1.2.6-dev`. So we might be able to get away with setting this value to `1.2.6-dev` and then replacing it with `1.3.0` _after_ the release candidate ships.

It's been a very long time since we've had a feature that was version-checked like this. So it's worth double-checking for yourself that old check we wrote handles cases where we have a suffix like this anyways, so that ex. `1.3.0+ent` users don't end up with a broken implementation. 😀

File: scheduler/reconcile_test.go
Comment (tgross) [inline scheduler/reconcile_test.go:348]: I would set all of these values to `true` in this test file so that we're not hiding any potential bugs. If you want to have a test that checks the code path, that's reasonable (maybe even tear it out after this release), but most tests should default to having the new always-on feature set to on.

File: scheduler/reconcile_util.go
Comment (tgross) [inline scheduler/reconcile_util.go:226]: After a cluster is fully upgraded we'll take this branch 100% of the time. It feels like the "optimization" of not allocating these empty maps for this narrow case is going to be outweighed by any possibility we screw it up and then end up with a panic on assignment to a nil map.

**Review by tgross — APPROVED** [review]
LGTM once my comment about the tests is addressed.

File: scheduler/reconcile_test.go
Comment (DerekStrickland) [reply scheduler/reconcile_test.go]: Fixed in latest push.

File: scheduler/reconcile_util.go
Comment (DerekStrickland) [reply scheduler/reconcile_util.go]: Fixed in latest push.

**Pull request merged**
